Source/Libraries/Adapters/Dnp3Adapters/Dnp3InputAdapter.cs

```diff
@@ -38,373 +39,423 @@
-            TimeSpan reconnectDelay = TimeSpan.FromMilliseconds(tcpConfig.reconnectDelayMs);
-            ChannelRetry channelRetry = new ChannelRetry(minRetry, maxRetry, reconnectDelay);
-            IChannelListener channelListener = new ChannelListener(state => OnStatusMessage(MessageLevel.Info, portName + " - Channel state change: " + state));
+            GetDataSource = () => DataSource,
+            MapQualityToStateFlags = MapQualityToStateFlags,
+            PublishFlagsAsSeparateMeasurements = PublishFlagsAsSeparateMeasurements,
+            TagMatchRegex = new Regex(TagMatchPattern, RegexOptions.Compiled),
+            QualityTagSuffix = QualityTagSuffix
+        };
 
-            IChannel channel = s_manager.AddTCPClient(portName, tcpConfig.level, channelRetry, new[] { new IPEndpoint(tcpConfig.address, tcpConfig.port) }, channelListener);
-            m_channel = channel;
+        m_soeHandler = new TimeSeriesSOEHandler(measurementLookup);
+        m_soeHandler.NewMeasurements += OnNewMeasurements;
 
-            IMaster master = channel.AddMaster(portName, m_soeHandler, DefaultMasterApplication.Instance, m_masterConfig.master);
+        // The TimestampDifferentiation property is a pass-through to the m_soeHandler.TimestampDifferentiation,
+        // so m_soeHandler must be initialized before reading this property from the connection string
+        if (settings.TryGetValue(nameof(TimestampDifferentiation), out setting) && double.TryParse(setting, out double timestampDifferentiation))
+            TimestampDifferentiation = timestampDifferentiation;
+        else
+            TimestampDifferentiation = DefaultTimestampDifferentiation;
 
-            if (m_pollingInterval > TimeSpan.Zero)
-                master.AddClassScan(ClassField.AllClasses, m_pollingInterval, m_soeHandler, TaskConfig.Default);
+        m_soeHandler.TimestampDifferentiation = TimeSpan.FromMilliseconds(TimestampDifferentiation);
 
-            master.Enable();
-            m_active = true;
-        }
+        if (settings.TryGetValue(nameof(PollingInterval), out setting) && double.TryParse(setting, out double pollingInterval))
+            PollingInterval = pollingInterval;
 
-        /// <summary>
-        /// Attempts to disconnect from data input source.
-        /// </summary>
-        /// <remarks>
-        /// Derived classes should attempt disconnect from data input source here.  Any exceptions thrown
-        /// by this implementation will be reported to host via <see cref="AdapterBase.ProcessException"/> event.
-        /// </remarks>
-        protected override void AttemptDisconnection()
-        {
-            if (!m_active)
-                return;
+        // Attach to output measurements for DNP3 device - just informs routing engine of expected measurements
+        OutputMeasurements = ParseOutputMeasurements(DataSource, false, $"FILTER ActiveMeasurements WHERE Device = '{Name}'");
```

Comment:
I realized this is what was causing the connect on demand thing to fail when nor measurements were unassigned, so that was the reasoning - this is basically what the ModBus adapter does. So, yes, could hide property, or we could alternately just default to this when there are not measurements defined.

Reply:
Of the two options, the latter seems better. It makes the adapter more configurable at the cost of nothing.

Reply:
https://github.com/GridProtectionAlliance/gsf/pull/271/commits/91ad023a01d10706cc5fb16f1b949367a13f9101